File: libs/python/mail_runtime_core/result_dispatch.py
#!/usr/bin/env python3
"""Shared ActionResult dispatch helpers."""

from __future__ import annotations

from typing import Callable

from .runtime import ActionResult


def dispatch_results(
    results: list[ActionResult],
    *,
    logger: Callable[[str], None],
    handlers: dict[str, Callable[[dict], None]] | None = None,
) -> None:
    """Deliver structured action results with service-provided side-effect handlers."""

    handlers = handlers or {}
    for result in results:
        if result.kind == "log":
            logger(result.payload["message"])
            continue

        handler = handlers.get(result.kind)
        if handler is None:
            logger(f"warn: unknown action result kind {result.kind}")
            continue

        handler(result.payload)

File: libs/ts/mail_runtime_core/src/result-dispatch.ts
/**
 * Shared ActionResult dispatch helpers.
 * TS port of mail_runtime_core/result_dispatch.py
 */

import type { ActionResult } from "./runtime.js";

export function dispatchResults(
  results: ActionResult[],
  options: {
    logger: (msg: string) => void;
    handlers?: Record<string, (payload: Record<string, unknown>) => void>;
  },
): void {
  const { logger, handlers = {} } = options;

  for (const result of results) {
    if (result.kind === "log") {
      logger(result.payload["message"] as string);
      continue;
    }

    const handler = handlers[result.kind];
    if (handler === undefined) {
      logger(`warn: unknown action result kind ${result.kind}`);
      continue;
    }

    handler(result.payload);
  }
}
